### PR TITLE
Added some additional improvements to the new RuleViolationException with causes

### PR DESCRIPTION
**Pull request by EricWittmann**
* No longer expose the "Difference" class too broadly.  
* Updated the Validity rule to allow returning causes (only supported by OpenApi and AsyncApi validators)

The `Difference` class is JSON Schema compatibility specific.  So I introduced a new `CompatibilityDifference` interface that we can use if we ever have support for fine-grained compatibility reporting for other artifact types.

Also now that we have "causes" available for rule violations, we can capture and return the semantic validation errors returned by the OpenAPI and AsyncAPI validators.

Hoping for some feedback from @jsenko and @ronsher and @adiohana on this. :)

### CODE REVIEW

**Comment by adiohana** [comment]
@EricWittmann 
this change makes sense and make the code more generic.
Is there any plan to support compatibility errors for other types as well? Is it depends on different teams/projects?

**Comment by EricWittmann** [comment]
Yes we aspire to support compatibility errors for all supported artifact types, however that is a lot of work.  We'll evolve that over time.

**Comment by adiohana** [comment]
got it.
Will be happy to join this effort.

File: app/src/main/java/io/apicurio/registry/rules/compatibility/GenericCompatibilityDifference.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    
+    /**
+     * @see io.apicurio.registry.rules.compatibility.CompatibilityDifference#asRuleViolationCause()
+     */
+    @Override
+    public RuleViolationCause asRuleViolationCause() {
+        RuleViolationCause cause = new RuleViolationCause();
+        cause.setDescription(message);
+        cause.setContext("/");
```
Comment (ronsher) [inline app/src/main/java/io/apicurio/registry/rules/compatibility/GenericCompatibilityDifference.java:51]: What does such context mean?

**Review by ronsher — APPROVED** [review]
LGTM

File: app/src/main/java/io/apicurio/registry/rules/compatibility/GenericCompatibilityDifference.java (hunk 1 above)
Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/rules/compatibility/GenericCompatibilityDifference.java:51]: The context is intended to mean where in the document the error/issue exists.  So basically a "path" into the document to help identify the location of the problem.  For most of our compatibility support we don't have that information.  For example, the Avro libraries don't provide a lot of feedback when doing this compatibility check - they give you a "yes" or "no".  So in these cases we just have a high level error message and  use "/" to indicate the root or entire document.

**Pull request merged**